### ISSUE TITLE
Improvements to `aria-labelledby`/`aria-describedby` support

### ISSUE THREAD

**Issue by ebruchez** [orbeon/orbeon-forms#5367]
#4832 provides initial support. This is about improvements to that support.

- [x] `<xf:help>` must also use `aria-describedby`, either to the help text or to instructions about how to get help
- [x] ~~add `role="group"` and add `aria-labelled-by`/`aria-described-by` as needed on XBL element~~
      - decided not to do because not useful
- [x] make sure "Attachment" and/or "Image Attachment" work, if not handle separately
- [x] remove existing uses of `aria-label` in XBL components
- [x] Q: `static-readonly` for `xf:input`, etc. should output `aria-labelledby`/`aria-describedby`?
    - Q: Yes, see #5525.
- [x] ~~add `xxbl:hint-for`, `xxbl:help-for`, `xxbl:alert-for`, maybe `xxbl:lhha-for`~~
    - not doing until we show that it's worth doing
- [ ] boolean `<xf:input>` must output `aria-describedby`/`aria-details` (but outputs `id`s now)
- [ ] `xf:trigger` must output `aria-describedby`/`aria-details` (but outputs `id` now for help)
    - 2022-10-25: There is a `title` on the elements, so probably find for the hint. What about help?
- [ ] automated testing
  - [ ] Q: can we write automated test for LHHA updates, iterations, etc.?

**Comment by ebruchez** [comment]
Make sure to support, for hint and help:

- `xf:input` with `type="xf:boolean"`
- `fr:number`
- `fr:currency`
- `fr:datetime`
- `xf:select1` with `appearance="search"`
- `fr:databound-select1` with `appearance="search"`
- `xf:select` with `appearance="full"`
- `fr:yesno-input`

[+1 from customer](https://3.basecamp.com/3600924/buckets/2016476/messages/3528991688)

**Comment by ebruchez** [comment]
Handling of `aria-details`:

- [ ] no longer output `aria-details`
- [ ] replace with `aria-described-by`
    - either point directly to the help text
    - or point to a generic text stating that there is further help available and how to obtain it

The first option is simpler, but the drawback is that a long text will be potentially read. This may not be a huge issue, depending on the type of help available. The spirit of the help, and of `aria-details`, is to provide much more detailed information, potentially very long. An indirection is definitely acceptable and, probably, better.

We can start with the first option, which is very simple, and then move to the second one.

**Comment by ebruchez** [comment]
For XBL components, currently nothing is done with `aria-` attributes when the binding doesn't have an `xxf:label-for` attribute and has non-custom ("standard") labels.

In this case, we should do what we suggest in the original comment:

- add `role="group"`
- use `aria-labelledby` and `aria-described-by`

**Comment by ebruchez** [comment]
Issue with `xxbl:label-for` when pointing to an HTML element, like in the case of `fr:tinymce`: the `<label>` produced has a `for` attribute, and the HTML element doesn't have an `aria-labelledby`.

The issue is that, in other cases, we point to an HTML [labelable](https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Content_categories#labelable), and so it's ok. But here we point to a `<div>`, and the relation should be reversed.

Two issues:

1.  How to figure out the element name? Do we even keep the `id` present?
2. We don't even keep the element in the XForms DOM.

We could use `xxf:control="true"` on the HTML element. Then, this should be the same as having an `<xf:group>` with `<xf:label for="...">`. However, currently, this doesn't work.

**Comment by ebruchez** [comment]
The case of an XBL component with `lhha` mode and no `xxbl:label-for` is fairly rare, in fact I am not sure we have an actual example. This said, my first implementation is placing attributes on the enclosing HTML element, and that is not good as that's around the LHHA elements as well. Instead, the attributes should be placed on an element placed around the body of the control markup.

However, all the content of the `<xbl:template>` element is used to produce the markup. So we can have multiple HTML elements output. There is not a single element we can use.

So maybe we simply need to require that there is an `xxbl:label-for` element.

**Comment by ebruchez** [comment]
Clarifications:

When we have an `xxbl:label-for`, we attach the `LHHAAnalysis` to two controls:

- the control to which the LHHA is attached by nesting or with `for` (`directTargetControl`)
- an optional nested control, if found, recursively following `xxbl:label-for`

Now, for the direct control, the `LhhaPlacementType` can be `Local` or `External`. This is orthogonal to whether there is an `xxbl:label-for` or not.

Right now, we incorrectly make a decision to produce a `for` attribute on the `<label>` vs. not an a `<span>`. To make a correct decision, we need to know whether the target is "labelable". Let's assume for now we use `xxf:control` to keep the name of the HTML element. Then we can assume, for example, that pointing to an `xf:group` as destination control is not labelable.

Next step: implement that logic.

**Comment by ebruchez** [comment]
The logic is now better. As proxy for a "labelable" element, we consider that we either:

- point to an XForms control which is a `CoreControl` but not an `AttributeControl`
- or point to an HTML element
    - here we assume we always point to a labelable HTML element
    - and if we need to point to a non-labelable element, use `xxf:control="true"` on the element

Next issue: if we use `xxf:control="true"`, this is represented by an `<xf:group>`. But `<xf:group>` does not produce `aria-*` attributes.

Next step: fix that.

**Comment by ebruchez** [comment]
So far I have modified XBL components to also have a `tabindex="0"` around groups. This has benefits and drawbacks. Voice Over in particular reads a group's `aria-*` attributes when focusing on a field inside the group. On the other hand, landing on the group itself has some benefits. Possibly, this could be an option.

Especially if we now use `xxf:control="true"` on labelled groups, the insertion of a `tabindex="0"` and `role="group"` could be easily centralized.

**Comment by ebruchez** [comment]
There is now a property to decide whether tabbing focuses on groups:

```xml
<property
    as="xs:boolean"
    name="oxf.xforms.a11y.focus-on-groups"
    value="true"/>
```

- [ ] Issue with focus on `fr:box-select` where the highlight is different from the rest:

<img width="245" alt="image" src="https://user-images.githubusercontent.com/105769/198109428-2c98c207-bf6e-4af4-a061-98106cdb8fa0.png">

- [ ] Issue with radio buttons showing a small box, which is inelegant:

<img width="249" alt="image" src="https://user-images.githubusercontent.com/105769/198110242-0ee0f652-d194-4d80-8ec8-a0cd0ba152c1.png">

**Comment by ebruchez** [comment]
- [ ] Highlights on Chrome are not all good-looking:

<img width="239" alt="image" src="https://user-images.githubusercontent.com/105769/198136165-63c4807e-0e46-41bf-84aa-5ed8eb13578b.png">

**Comment by ebruchez** [comment]
- [ ] `fr:date` doesn't read hint

**Comment by ebruchez** [comment]
For `fr:date`: we have a `<label for="...">`, which works, and a `<span class="xforms-hint">`, which is ok, but the hint is not linked with `aria-described-by`.

**Comment by ebruchez** [comment]
Issue is we have an `<xh:input>`, and we don't have a handler for it. We could use `xxf:control="true"`, but then we wouldn't it to have a group, although we could use `role="textbox"`.

**Comment by ebruchez** [comment]
Issue for `fr:date`. We have:

```xml
<xh:input type="text" placeholder="{$placeholder}" id="input"/>
```

But this means the hint is not read, as the element is a plain HTML element and there is no hint handling for those right now. If we make the element an `xxf:control` with:

```xml
<xh:input xh:type="text" placeholder="{$placeholder}" id="input" xxf:control="true" role="textbox"/>
```

Then the `placeholder` attribute is ignored. If we use `xh:placeholder`, it is copied, but the AVT is not resolved.

**Comment by ebruchez** [comment]
One of the obvious remaining issues is that the hint is not read on some controls that either:

- use `xxbl:label-for`
- built-in checkboxes/radio buttons groups

For the former, I think that this might be due to the fact that we only handle, well, labels, with `xxbl:label-for`. Hence the suggestion above to add "`xxbl:hint-for`, `xxbl:help-for`, `xxbl:alert-for`, maybe `xxbl:lhha-for`". Possibilities:

1. handle LHHA separately add add individual `xxbl:hint-for`, `xxbl:help-for`, `xxbl:alert-for`
    - Q: Is there a use case?
1. assume that `xxbl:label-for` implies all the other LHHA, and then consider it an alias for `xxbl:lhha-for` for backward compatibility
1. `xxbl:lhha-for` first and update components

**Comment by ebruchez** [comment]
In `PartAnalysisSupport`, it seems that we handle all LHHA with `xxbl:label-for`

**Comment by ebruchez** [comment]
For the Dynamic Dropdown, we do generate an `aria-describedby` on the `select`, but that select is hidden so maybe that's why it's not read? However, the label is read, somehow.

For the checkboxes, it's an easy fix.

For `fr:number`/`fr:currency`, we produce an incorrect `aria-describedby="fr-view-component≡typed-controls-section≡grid-3-grid≡number-control≡suffix"`, when the hint has `id="fr-view-component≡typed-controls-section≡grid-3-grid≡number-control≡≡t"`.

**Comment by ebruchez** [comment]
Ah, for `fr:number`/`fr:currency`, the `xh:input` has an explicit `aria-describedby="{string-join((xxf:client-id('prefix'), xxf:client-id('suffix')), ' ')}"`. The label works because it can use a `for`, but the hint doesn't work because we don't automatically generate `aria-describedby` on plain HTML elements, and that's why we use `xxf:control="true"` instead.

**Comment by ebruchez** [comment]
The problem, if we use `xxf:control="true"` like for `fr:date`, is that HTML AVTs are not evaluated. For example:

- `xh:step="any"` is not an AVT and shows as `step="any"` which is ok
- but `xh:pattern`, `xh:min`, `xh:maxlength`, and the custom `aria-describedby` are not handled

**Comment by ebruchez** [comment]
We have `ControlExtensionAttributesSupport` to handle custom/extension attributes. This is how this works:

- static controls have `extensionAttributes`, which includes
    - `style`, `class`, and `role`
    - `data-*`
    - `allowedExtensionAttributes` for each control
        - `xf:secret`
            - `xxf:size`
            - `xxf:maxlength`
            - `xxf:autocomplete`
        - `xf:output[mediatype = 'image/*']`
            - `alt`
        - `xf:output[appearance = 'xxf:download']`
            - `xxf:target`
        - `xf:textarea`
            - `xxf:maxlength`
            - `xxf:cols`
            - `xxf:rows`
        - `xf:input`
            - `xxf:size`
            - `xxf:title`
            - `xxf:maxlength`
            - `xxf:pattern`
            - `xxf:autocomplete`
        - selection
            - if radio, `xxf:group`
            - `xxf:title`
        - `xf:upload`
            - `accept`
            - `mediatype`
            - `xxf:title`
        - `xf:group`
            - if `td` or `tr`, `colspan` and `rowspan`
        - `xf:trigger[appearance = minimal]`
            - `xxf:title`
- values evaluated as needed
- compared for Ajax
- `addExtensionAttributesExceptClassAndAcceptForHandler` for handlers
    - called by `XFormsBaseHandlerXHTML.getContainerAttributes` for empty namespace
    - called by several handlers for `xxf:` namespace
- `addExtensionAttributesExceptClassAndAcceptForAjax` for Ajax
    - called by `outputAjaxDiff`
    - only handling attributs in empty namespace?

So if we wanted to support `xh:` attributes on groups, what would it take?

**Comment by ebruchez** [comment]
Internally, `xxf:control="true"` is handled by a group (which is not great). But we could say that if the group is for example backing an `xh:input`, then the `pattern`, `min`, and `maxlength` could be handled.

For the custom `arai-describedby`, this could be handled more generally (space-separated ids).

**Comment by ebruchez** [comment]
Noting that XHTML attributes are just copied over but not evaluated as AVTs.

**Comment by ebruchez** [comment]
The the Single Checkbox, the hint is currently not linked. The label is custom and set as the item label, and the help is copied inside. But the hint remains "outside" and unlinked.

I thought we could make the hint custom as well, and use an item hint, but item hints are not linked with `aria-describedby` right now (which is another a11y issue).

We also don't support `xxbl:label-for` (or `xxbl:hint-for`) to point to an item.

We could have a `xxbl:label-for` (or `xxbl:hint-for`) pointing to the nested `<xf:select1>`. This however would translate into a group around the checkbox, and that would be what points to the hint, which wouldn't be ideal.

Having an item hint would be better, probably. Then the hint would look like an underline, but would need an `aria-labelledby`.

**Comment by ebruchez** [comment]
Round 2 covered by #5570.
